### PR TITLE
fix: ignore invalid app_ids in /update and /stats requests

### PR DESCRIPTION
**Pull request by ayewo**
Fix to /claim https://github.com/Cap-go/capgo/issues/384

### CODE REVIEW

**Comment by riderx** [comment]
Hey @ayewo thanks a lot for the fix, my ask was to use zod package to validate the fields, and no only the appId but all fields

**Comment by WcaleNieWolny** [comment]
I do not think you can validate 
all fields. I do not see a way to implement that. Validating only app is is okay as far as I am concerned.

The only things are missing are tests for this. @ayewo Can you please add some tests for this? (See the latest merged PR with backend tests)

This test should test the following edge cases:
 - Empty app name
 - weird app name (like the one Martin shared)
 - a very long app id (that is valid but has 30 dots for example)

File: supabase/functions/stats/index.ts [hunk 1]
```diff
@@ -13,9 +14,54 @@ const failActions = [
...
+const jsonRequestSchema = z.object({
+  app_id: z.string({
+    required_error: "App ID is required",
+    invalid_type_error: "App ID name must be a string",
+  }),
+  device_id: z.string({
+    required_error: "Device ID is required",
+    invalid_type_error: "Device ID must be a string",
+  }).max(36),
+  platform: z.string({
```
Comment (WcaleNieWolny) [inline supabase/functions/stats/index.ts]: A platform could a union perhaps?

File: supabase/functions/stats/index.ts [hunk 2]
```diff
@@ -13,9 +14,54 @@ const failActions = [
...
+    invalid_type_error: "Version OS must be a string",
+  }),
+  version_code: z.optional(z.string()),
+  version_build: z.optional(z.string()),
+  action: z.optional(z.string()),
+  custom_id: z.optional(z.string()),
+  channel: z.optional(z.string()),
+  plugin_version: z.optional(z.string()),
+  is_emulator: z.optional(z.boolean()),
+  is_prod: z.optional(z.boolean()),
```
Comment (WcaleNieWolny) [inline supabase/functions/stats/index.ts]: Can we use the default values? Like is_emulator the default would be false and is prod would be by default true

File: supabase/functions/updates/index.ts [hunk 3]
```diff
@@ -3,20 +3,34 @@ import { serve } from 'https://deno.land/std@0.200.0/http/server.ts'
...
+    required_error: "Device ID is required",
+    invalid_type_error: "Device ID must be a string",
+  }).max(36),
+  version_name: z.string({
+    required_error: "Version name is required",
+    invalid_type_error: "Version name must be a string",
+  }),
   is_emulator: z.optional(z.boolean()),
-  is_prod: z.optional(z.boolean()),
+  is_prod: z.optional(z.boolean())
```
Comment (WcaleNieWolny) [inline supabase/functions/updates/index.ts]: I made a mistake while working on this zod schema, if you want you can port the fixed version from [here](https://github.com/Cap-go/capgo/pull/387) or wait until that PR is merged and resolve merge conflicts

File: supabase/functions/updates/index.ts [hunk 4]
```diff
@@ -26,6 +40,13 @@ const headersSchema = z.object({
 const bypassRedis = false
 
 async function main(url: URL, headers: BaseHeaders, method: string, body: AppInfos) {
+  let parseResult = null
+  try {
+      parseResult = jsonRequestSchema.parse(body)
```
Comment (WcaleNieWolny) [inline supabase/functions/updates/index.ts]: There is something called a “safeParse” you can use that instead of a try catch block

File: supabase/functions/stats/index.ts [hunk 5]
```diff
@@ -13,9 +14,54 @@ const failActions = [
...
+}).refine((data) => deviceIdRegex.test(data.device_id), {
+  message: "Device ID must be a valid UUID string"
+});
+
 async function main(url: URL, headers: BaseHeaders, method: string, body: AppStats) {
   try {
     console.log('body', body)
+    let parseResult = null
+    try {
+        parseResult = jsonRequestSchema.parse(body)
```
Comment (WcaleNieWolny) [inline supabase/functions/stats/index.ts]: Perhaps use zod safe parse, there should exist a method that does not throw but instead returns the error

**Review by WcaleNieWolny — CHANGES_REQUESTED** [review]
Please resolve the comments I left above, but over all the changes are a big improvement over what we had

File: supabase/functions/stats/index.ts (hunk 1 above)
Comment (ayewo) [reply supabase/functions/stats/index.ts]: When I checked now, nope since it's used as a container for mixing types (e.g string and number).

**Comment by ayewo** [comment]
@WcaleNieWolny thanks for your comments.

I've pulled in the latest from `main` by discarding my commits, but that ended up closing this PR so I opened a fresh PR in #399.

File: supabase/functions/stats/index.ts (hunk 5 above)
Comment (ayewo) [reply supabase/functions/stats/index.ts]: In my testing, `safeParse` is lenient. It will let incorrectly formatted fields through.

it doesn't flag invalid JSON fields unless you wrap `parse` inside a `try / catch`.

File: supabase/functions/updates/index.ts (hunk 4 above)
Comment (ayewo) [reply supabase/functions/updates/index.ts]: See my explanation regarding `safeParse` below.
